File: src/lua/fpu/fft.lua
FFT = class(CrystalBenchmarkBase, function(fft, writer)
    CrystalBenchmarkBase.init(fft, writer, "FFT")
end)

FFT.N = 128

function FFT:benchImplementation()
    local count = 0
    local PI = math.pi

    local k, i, j = 0, 0, 0
    local w, wr, wi, x, xr, xi = 0.0, 0.0, 0.0, 0.0, 0.0, 0.0

    local fr = {}
    local fi = {}

    for ix = 1, FFT.N, 1 do
        fr[ix] = 0
        fi[ix] = 0
    end

    while not self:isCancelled() do
        for ix = 1, FFT.N, 1 do
            x = -2 * PI + ix * 4.0 * PI / FFT.N
            fr[ix] = math.sin(x) + math.sin(2.0 * x)
            fi[ix] = 0.0
        end

        fr = self:bitreverse(fr, FFT.N)
        fi = self:bitreverse(fi, FFT.N)

        k = 1
        while k < FFT.N do
            j = 1
            while j <= k do
                w = 2.0 * PI / (2.0 * k)
                wr = math.cos(w * j)
                wi = -math.sin(w * j)

                i = j
                while i <= FFT.N do
                    xr = wr * fr[i + k] - wi * fi[i + k]
                    xi = wr * fi[i + k] + wi * fr[i + k]
                    fr[i + k] = fr[i] - xr
                    fi[i + k] = fi[i] - xi
                    fr[i] = fr[i] + xr
                    fi[i] = fi[i] + xi

                    i = i + (k * 2)
                end
                j = j + 1
            end
            k = k * 2
        end

        count = count + 1
    end

    return (count * FFT.N) / 1153.3980
end

function FFT:bitreverse(a, N)
    local j, k, l = 0, 0, 0
    local b = {}
    for i = 1, N do
        b[i] = 0
    end

    for i = 1, N, 1 do
        k = 1
        l = math.floor(N / 2)

        j = 1
        while j < N do
            if bitand(i, j) == 1 then
                k = k + l
            end
            l = math.floor(l / 2)
            j = j * 2
        end

        b[i] = a[k];
    end

    for i = 1, N, 1 do
        a[i] = b[i]
    end

    return a
end

File: src/dotnet/EntityFX.CrystalMarkNet/FPU/FFT.cs
using System;
using System.Threading;

namespace EntityFX.CrystalMarkNet.FPU
{
    class FFT : CrystalBenchmarkBase
    {
        private const int N = 128;
        protected override double BenchImplementation(CancellationToken cancellationToken)
        {
            ushort count = 0;
            const float PI = (float)Math.PI;

            int i, j, k;

            float[] fr = new float[N];
            float[] fi = new float[N];

            float w, wr, wi, x, xr, xi;

            for (; ; )
            {
                for (i = 0; i < N; i++)
                {
                    x = -2 * PI + (float)i * 4.0f * PI / (float)N;
                    fr[i] = (float)(Math.Sin(x) + Math.Sin(2.0f * x));
                    fi[i] = 0.0f;
                }

                Bitreverse(ref fr, N);
                Bitreverse(ref fi, N);

                for (k = 1; k < N; k = k * 2)
                {
                    for (j = 0; j < k; j++)
                    {
                        w = 2.0f * (float)PI / (2.0f * (float)k);
                        wr = (float)Math.Cos(w * j);     
                        wi = (float)-Math.Sin(w * j);
                        for (i = j; i < N; i += (k * 2))
                        {
                            xr = wr * fr[i + k] - wi * fi[i + k];
                            xi = wr * fi[i + k] + wi * fr[i + k];
                            fr[i + k] = fr[i] - xr;
                            fi[i + k] = fi[i] - xi;
                            fr[i] = fr[i] + xr;
                            fi[i] = fi[i] + xi;
                        }
                    }
                }
                count++;

                if (cancellationToken.IsCancellationRequested)
                {
                    return (double)count * N / 1153.3980d * 25;
                }
            }
        }

        private static void Bitreverse(ref float[] a, int N)
        {
            int i, j, k, l;
            float[] b = new float[N];

            for (i = 0; i < N; i++)
            {
                k = 0;
                l = N / 2;
                for (j = 1; j < N; j *= 2)
                {
                    if ((i & j) == 1) k += l;
                    l /= 2;
                }
                b[i] = a[k];
            }
            for (i = 0; i < N; i++)
            {
                a[i] = b[i];
            }
        }
    }
}